x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/persistence/DataFrameTransformsConfigManager.java

```diff
@@ -110,6 +121,51 @@ public void getTransformConfiguration(String transformId, ActionListener<DataFra
         }));
     }
 
+    /**
+     * Get more than one DataFrameTransformConfig
+     *
+     * @param transformId Can be a single transformId, `*`, or `_all`
+     * @param resultListener Listener to alert when request is completed
+     */
+    public void getTransformConfigurations(String transformId,
+                                           ActionListener<List<DataFrameTransformConfig>> resultListener) {
+        BoolQueryBuilder queryBuilder = QueryBuilders.boolQuery()
+            .filter(QueryBuilders.termQuery("doc_type", DataFrameTransformConfig.NAME));
+        if (Strings.isAllOrWildcard(new String[]{transformId}) == false) {
+            queryBuilder.filter(QueryBuilders.termQuery(DataFrameField.ID.getPreferredName(), transformId));
+        }
+
+        SearchRequest request = client.prepareSearch(DataFrameInternalIndex.INDEX_NAME)
+            .setIndicesOptions(IndicesOptions.LENIENT_EXPAND_OPEN)
+            .setTrackTotalHits(true)
+            .setSize(10_000) // TODO support pagination
+            .setQuery(queryBuilder)
+            .request();
+
+        ClientHelper.executeAsyncWithOrigin(client.threadPool().getThreadContext(), DATA_FRAME_ORIGIN, request,
+            ActionListener.<SearchResponse>wrap(
+                searchResponse -> {
+                    List<DataFrameTransformConfig> configs = new ArrayList<>(searchResponse.getHits().getHits().length);
+                    for (SearchHit hit : searchResponse.getHits().getHits()) {
+                        DataFrameTransformConfig config = parseTransformLenientlyFromSourceSync(hit.getSourceRef(),
```

Comment:
I am worried: what's the size of 10000 config objects? I would not be worried about the `XContent`, but we parse all of those and create objects including nested objects like AggregationBuilders, QueryBuilders, etc. In addition: for `GET` we parse every query but do not use it.

What about having different methods? In the end we have 2 usecases:

- getting exactly 1 config for the purpose of using it (indexer)
- getting the configuration of 1 or more config for the purpose of returning the XContent of it (`GET`), if possible avoid object creations but at least do not keep 10k objects in a list

Reply:
Making this change would make data frame transforms different to every other endpoint in the code base.

Remember also that in the transport client the `GET` endpoint needs to return valid objects, not XContent.

If 10000 is too big then the solution that fits with the way everything else works would be to implement pagination and restrict the maximum page size to something less than 10000.  But I think that should be done in a different PR (if at all).

Reply:
I am adding pagination to this endpoint (and stats) eventually and am planning on doing that in a separate PR.

Reply:
@droberts195 @benwtrent I am unsure if you understood my point. This wasn't about endpoints but about the _methods_ in this class. Imagine we really have 10k transforms, we are creating a list of 10k _objects_, those objects have inner objects. These 10k "rich objects" are temporary objects as you call `toXContent()` on them at the end. Only the string representation is needed, the parsed objects are just side artifacts.

It's correct that the response object has to support wire serialization, however the inner implementation of the response object is completely up to us. There is no need to store a list of `DataFrameTransformConfig` objects. The requirements of a response object can be fulfilled without storing these "rich objects"[*]. 

Pagination doesn't solve the problem as long as we keep the size of a single page at 10k (see line 142).

[*] Having that said, I simply do not see a need to return 10k configs at a time. I do not even see the need to solve the >10k transforms usecase in 7.1.

I am fine with "fixing this in an upcoming PR" - but I strongly disagree doing it this way. At line 142 we explicitly set the limit to 10k and as explained above I do not think the implementation is ready to support 10k due to the high memory usage it requires. We should set a lower limit there for the time being and we can "fix this in an upcoming PR". We should not put the system at risk for this! Progress over perfection but in a safe way.

Pagination in combination with a smaller page size sounds like a good alternative, again getting 10k configs at a time, I do not see a need for it.

Reply:
@hendrikmuhs when searching and getting a result, we will indicate the size and page parameters given (after another refactor to move common code in to xpack.core, see https://github.com/elastic/elasticsearch/pull/39976). These paging params will default to `size 100` and `start: 0`, respectively. 

Addendum, I am against passing around strings and maps more than necessary. This sacrifices one of the securities of having a statically typed language. I think worrying about this type of serialization overhead is pre-optimization at the cost of compile time guarantees.

I am fine lowering the default in this PR for now, but I think chances of this bringing down a cluster due to somebody creating 10k transforms between us merging this PR and then putting in sensible paging parameters in a follow up PR are infinitesimal (7.1 is not even released yet.).

Reply:
I am fine if that gets addressed in a PR that hits 7.1, sure. When the discussion started it wasn't clear what "follow up PR" means (also given that the problem existed prior this PR).

Pagination with a sensible limit per page sounds great! No need to further discuss about optimizing the size of the list once we have that.

Reply:
We need to step away from the idea of returning > 10K results. I think the way forward is to allow searching (i.e. implement the `AbstractTransportGetResourcesAction`). Then if a user hits > 10K results, we should error. The user can then change the query in order to reduce the matching resources.